### ISSUE TITLE
Indexing errors when using patched Swift package with `rules_xcodeproj`

### ISSUE THREAD

**Issue by cgrindel** [cgrindel/rules_swift_package_manager#409]
So `bazel build` works without issues but noticing an issue with `rules_xcodeproj` (I imagine because it enables index while building feature):

This is an index error so it doesn't stop/fail the build but does show an error on the `import Markdown` line

```sh
/Applications/Xcode-14.2.0.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swiftc -module-name BazelPostflight -Onone -enforce-exclusivity=checked -Xcc -working-directory=/var/tmp/_bazel_lpadron/9ae9416857eb79bb978de35a53d54970/rules_xcodeproj.noindex/build_output_base/execroot/cash-ios -working-directory=/var/tmp/_bazel_lpadron/9ae9416857eb79bb978de35a53d54970/rules_xcodeproj.noindex/build_output_base/execroot/cash-ios -Xcc -ivfsoverlay -Xcc /Users/lpadron/Library/Developer/Xcode/DerivedData/CashCLI-gpdwzxbdldztlmdeuxjtnhgkfwyz/Index.noindex/Build/Intermediates.noindex/bazel-out-overlay.yaml -vfsoverlay /Users/lpadron/Library/Developer/Xcode/DerivedData/CashCLI-gpdwzxbdldztlmdeuxjtnhgkfwyz/Index.noindex/Build/Intermediates.noindex/bazel-out-overlay.yaml @/Users/lpadron/Library/Developer/Xcode/DerivedData/CashCLI-gpdwzxbdldztlmdeuxjtnhgkfwyz/Index.noindex/Build/Intermediates.noindex/CashCLI.build/bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/Tools/CashCLI/BazelPostflight/DerivedSources/swift.compile.params -sdk /Applications/Xcode-14.2.0.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -target arm64-apple-macos12.0 -enable-bare-slash-regex -g -module-cache-path /Users/lpadron/Library/Developer/Xcode/DerivedData/ModuleCache.noindex -Xfrontend -serialize-debugging-options -enable-testing -swift-version 5 -I bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/external/swiftpkg_swift_argument_parser -I bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/external/swiftpkg_swift_collections -I bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/external/swiftpkg_swift_log -I bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/Tools/CashCLI -I bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/external/swiftpkg_swift_markdown -Xcc -Xclang -Xcc -fmodule-format=raw -Xcc -Xclang -Xcc -detailed-preprocessing-record -num-threads 10 -emit-module -emit-module-path /Users/lpadron/Library/Developer/Xcode/DerivedData/CashCLI-gpdwzxbdldztlmdeuxjtnhgkfwyz/Index.noindex/Build/Intermediates.noindex/CashCLI.build/bazel-out/darwin_arm64-dbg-ST-3182f29850cb/bin/Tools/CashCLI/BazelPostflight/Objects-normal/arm64/BazelPostflight.swiftmodule -Xcc -working-directory/var/tmp/_bazel_lpadron/9ae9416857eb79bb978de35a53d54970/rules_xcodeproj.noindex/build_output_base/execroot/cash-ios /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/Actions/CIClearBazelFailureCommentsAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/Actions/CICollectBazelLogsAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/Actions/CICollectBazelSnapshotTestOutputsAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/Actions/CICommentBazelFailuresAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/Actions/CleanStaleBazelCacheAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/Actions/CollectBazelSnapshotTestOutputsAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/PostflightAction.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/PostflightActionConfig.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/PostflightActionError.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Action/PostflightActionRunner.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/BazelPostflightCommand.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/GitHub/GitHubService+Postflight.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/LogViolationParser.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/LogViolationSummary.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Model/BEPEvent+Extensions.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Model/BEPEvent+Parse.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Model/BEPEvent.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Model/BEPEventID.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Model/CmdLineOption.swift /Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/Model/LogViolation.swift

/Users/lpadron/Development/cash-ios/Tools/CashCLI/Sources/BazelPostflight/LogViolationSummary.swift:3:8: error: missing required modules: 'CAtomic', 'cmark_gfm', 'cmark_gfm_extensions'
```

Could it be the names are incorrect here?
<img width="377" alt="Screenshot 2023-06-08 at 1 54 39 PM" src="https://github.com/cgrindel/rules_swift_package_manager/assets/13840545/1373bf74-b863-4baf-8b24-25741f465c3f">

_Originally posted by @luispadron in https://github.com/cgrindel/rules_swift_package_manager/issues/400#issuecomment-1583096735_

**Comment by cgrindel** [comment]
> Could it be the names are incorrect here?

@luispadron Below is the target that is generated.

```python
cc_library(
    name = "Sources_CAtomic",
    copts = [
        "-fblocks",
        "-fobjc-arc",
        "-fPIC",
        "-fmodule-name=CAtomic",     // <== clang module name here
    ],
    defines = ["SWIFT_PACKAGE=1"],
    deps = [],
    hdrs = ["Sources/CAtomic/include/CAtomic.h"],
    includes = ["Sources/CAtomic/include"],
    srcs = ["Sources/CAtomic/CAtomic.c"],
    tags = ["swift_module=CAtomic"],    // <== Swift module name here
    textual_hdrs = ["Sources/CAtomic/CAtomic.c"],
    visibility = ["//visibility:public"],
)
```

@brentleyjones For `cc_library` targets, does `rules_xcodeproj` expect the target name to be the module name?

**Comment by cgrindel** [comment]
@luispadron I just created an Xcode project using`rules_xcodeproj` in the `examples/ios_sim` workspace. I am not seeing any indexing errors. Xcode is autocompleting attributes on the `Document` object from `Markdown`. Can you see if this works for you?

Branch: https://github.com/cgrindel/rules_swift_package_manager/tree/debug_rules_xcodeproj_error
Xcode Version: 14.2

**Comment by luispadron** [comment]
Not seeing the errors either on that branch, I'm guessing something in our `.bazelrc` is causing this. Let me try and narrow it down.

**Comment by luispadron** [comment]
I'm able to repro, in our Cash App project, the index failure by toggling:

```sh
build --macos_minimum_os=12.0 # FAIL
build --macos_minimum_os=13.0 # OK
```

Adding this and moving `Markdown` to the source vs. test for your example branch seems to fail with a similar module not found (bot for `NIO`), here is the index log:

```
Loading:
WARNING: For repository 'cgrindel_bazel_starlib', the root module requires module version cgrindel_bazel_starlib@0.14.9, but got cgrindel_bazel_starlib@0.15.0 in the resolved dependency graph.
Loading:
Loading: 0 packages loaded
Analyzing: target @rules_xcodeproj~1.7.0~internal~rules_xcodeproj_generated//generator/xcodeproj:xcodeproj (0 packages loaded, 0 targets configured)
Analyzing: target @rules_xcodeproj~1.7.0~internal~rules_xcodeproj_generated//generator/xcodeproj:xcodeproj (0 packages loaded, 0 targets configured)
INFO: Repository rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio instantiated at:
  callstack not available
Repository rule swift_package defined at:
  /private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/swift_package.bzl:165:32: in <toplevel>
ERROR: An error occurred during the fetch of repository 'rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio':
   Traceback (most recent call last):
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/swift_package.bzl", line 65, column 28, in _swift_package_impl
		pkg_ctx = pkg_ctxs.read(repository_ctx, directory, env)
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/pkg_ctxs.bzl", line 13, column 28, in _read
		pkg_info = pkginfos.get(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/pkginfos.bzl", line 75, column 39, in _get
		dump_manifest = _get_dump_manifest(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/pkginfos.bzl", line 32, column 57, in _get_dump_manifest
		return repository_utils.parsed_json_from_spm_command(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/repository_utils.bzl", line 67, column 49, in _parsed_json_from_spm_command
		json_str = repository_utils.exec_spm_command(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/repository_utils.bzl", line 54, column 13, in _execute_spm_command
		fail(err_msg_tpl.format(
Error in fail: Failed to execute SPM command. name: rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio, args: ["xcrun", "swift", "package", "dump-package"]
.
ERROR: <builtin>: fetching swift_package rule @rules_swift_package_manager~override//:rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio: Traceback (most recent call last):
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/swift_package.bzl", line 65, column 28, in _swift_package_impl
		pkg_ctx = pkg_ctxs.read(repository_ctx, directory, env)
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/pkg_ctxs.bzl", line 13, column 28, in _read
		pkg_info = pkginfos.get(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/pkginfos.bzl", line 75, column 39, in _get
		dump_manifest = _get_dump_manifest(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/pkginfos.bzl", line 32, column 57, in _get_dump_manifest
		return repository_utils.parsed_json_from_spm_command(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/repository_utils.bzl", line 67, column 49, in _parsed_json_from_spm_command
		json_str = repository_utils.exec_spm_command(
	File "/private/var/tmp/_bazel_lpadron/b5a02afc288b0797cc26043c5d3c3c9c/rules_xcodeproj.noindex/indexbuild_output_base/external/rules_swift_package_manager~override/swiftpkg/internal/repository_utils.bzl", line 54, column 13, in _execute_spm_command
		fail(err_msg_tpl.format(
Error in fail: Failed to execute SPM command. name: rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio, args: ["xcrun", "swift", "package", "dump-package"]
.
ERROR: /Users/lpadron/Development/rules_swift_package_manager/examples/ios_sim/Tests/FooTests/BUILD.bazel:4:14: //Tests/FooTests:FooTestsLib depends on @rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio//:Sources_NIO in repository @rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio which failed to fetch. no such package '@rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio//': Failed to execute SPM command. name: rules_swift_package_manager~override~swift_deps~swiftpkg_swift_nio, args: ["xcrun", "swift", "package", "dump-package"]
.
ERROR: Analysis of target '@rules_xcodeproj~1.7.0~internal~rules_xcodeproj_generated//generator/xcodeproj:xcodeproj' failed; build aborted:
INFO: Elapsed time: 1.479s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
Command PhaseScriptExecution failed with a nonzero exit code
```

I pushed up a commit with this reproducible if you build `Foo` target in Xcode explicitly

**Comment by luispadron** [comment]
So steps:

- `bazel run //:xcodeproj`
- Open `App.xcodeproj`
- `Select > Foo` target in Xcode
- Go to `Bar.swift`
- Build
- Wait for index failure

Might need to use `defaults write com.apple.dt.Xcode IDEIndexShowLog 1` to see index logs

**Comment by luispadron** [comment]
I found that I don't need to use the bazelrc changes. Moving the dep from the test target to source target in the example was enough to produce the NIO indexing error. Updated the branch

**Comment by cgrindel** [comment]
Just a quick update. Yesterday, I was able to reproduce the indexing error. From the Xcode logs, I identified the failing `swiftc` execution. I added `-v` to that invocation and see the failing `swift-frontend` invocation. I spent some time trying to coax `swift-frontend` to provide more information (e.g. logging) without much success. 😢

On the chance that the utility was not finding the `CNIOAtomic` module due to the name of the modulemap, I manually copied the version that was generated to be named `CNIOAtomic.swift.modulemap`. This did not change the failure. So, I am reasonably certain that is not the issue.

Not having worked much with `rules_xcodeproj` lately, I will create a copy of the `ios_sim` example without the patches to see if I can reproduce the indexing error. Mainly, I would like to understand if the indexing issues are related to the Swift package patch. Also, I will connect with @brentleyjones to glean any insights on how to debug this issue.

**Comment by cgrindel** [comment]
I am able to reproduce the issue without using the new patch mechanism. The original `ios_sim` example experiences a `no such module `NIO`` indexing error.

```
/Users/chuck/code/cgrindel/rules_swift_package_manager/debug_rules_xcodeproj_error/examples/ios_sim_no_patch/Sources/Foo/Bar.swift:1:8: error: no such module 'NIO'
import NIO
       ^
```

**Comment by cgrindel** [comment]
@luispadron It looks like we can work around the issue by executing the following before launching the Xcode project:

```sh
$ bazel run //:xcodeproj -- --config=indexbuild --generator_output_groups=all_targets build
```

I would think that something in `rules_xcodeproj` should do this automatically.

**Comment by cgrindel** [comment]
[Related Slack thread](https://bazelbuild.slack.com/archives/C0429BN0F8V/p1686606149856739?thread_ts=1686603253.256269&cid=C0429BN0F8V)

**Comment by luispadron** [comment]
Nice find! Thanks for digging into it, I'll verify that fixes it on our project as well tomorrow morning

**Comment by cgrindel** [comment]
@luispadron Did you have a chance to try out the workaround? Also, given that this seems to be related to how `rules_xcodeproj` and Xcode interact, I am inclined to close this issue. Thoughts?

**Comment by luispadron** [comment]
I haven't but plan to take a look soon! And yes that makes sense to me
